src/nvim/eval/typval.c

```diff
@@ -0,0 +1,2560 @@
+void tv_list_watch_fix(list_T *const l, const listitem_T *const item)
+  FUNC_ATTR_NONNULL_ALL
+{
+  for (listwatch_T *lw = l->lv_watch; lw != NULL; lw = lw->lw_next) {
+    if (lw->lw_item == item) {
+      lw->lw_item = item->li_next;
+    }
+  }
+}
+
+//{{{2 Alloc/free
+
+/// Allocate an empty list
+///
+/// Caller should take care of the reference count.
+///
+/// @return [allocated] new list.
+list_T *tv_list_alloc(void)
+  FUNC_ATTR_NONNULL_RET FUNC_ATTR_MALLOC
+{
+  list_T *const list = xcalloc(1, sizeof(list_T));
+
+  // Prepend the list to the list of lists for garbage collection.
+  if (gc_first_list != NULL) {
+    gc_first_list->lv_used_prev = list;
+  }
+  list->lv_used_prev = NULL;
+  list->lv_used_next = gc_first_list;
+  gc_first_list = list;
+  return list;
+}
+
+/// Free items contained in a list
+///
+/// @param[in,out]  l  List to clear.
+void tv_list_free_contents(list_T *const l)
+  FUNC_ATTR_NONNULL_ALL
+{
+  for (listitem_T *item = l->lv_first; item != NULL; item = l->lv_first) {
+    // Remove the item before deleting it.
+    l->lv_first = item->li_next;
+    tv_clear(&item->li_tv);
+    xfree(item);
+  }
+  l->lv_len = 0;
+  l->lv_idx_item = NULL;
+  for (listwatch_T *lw = l->lv_watch; lw != NULL; lw = lw->lw_next) {
+    lw->lw_item = NULL;
+  }
+}
+
+/// Free a list itself, ignoring items it contains
+///
+/// Ignores the reference count.
+///
+/// @param[in,out]  l  List to free.
+void tv_list_free_list(list_T *const l)
+  FUNC_ATTR_NONNULL_ALL
+{
+  // Remove the list from the list of lists for garbage collection.
+  if (l->lv_used_prev == NULL) {
+    gc_first_list = l->lv_used_next;
+  } else {
+    l->lv_used_prev->lv_used_next = l->lv_used_next;
+  }
+  if (l->lv_used_next != NULL) {
+    l->lv_used_next->lv_used_prev = l->lv_used_prev;
+  }
+
+  xfree(l);
+}
+
+/// Free a list, including all items it points to
+///
+/// Ignores the reference count. Does not do anything if
+/// tv_in_free_unref_items is true.
+///
+/// @param[in,out]  l  List to free.
+void tv_list_free(list_T *const l)
+  FUNC_ATTR_NONNULL_ALL
+{
+  if (!tv_in_free_unref_items) {
+    tv_list_free_contents(l);
+    tv_list_free_list(l);
+  }
+}
+
+/// Unreference a list
+///
+/// Decrements the reference count and frees when it becomes zero or less.
+///
+/// @param[in,out]  l  List to unreference.
+void tv_list_unref(list_T *const l)
+{
+  if (l != NULL && --l->lv_refcount <= 0) {
+    tv_list_free(l);
+  }
+}
+
+//{{{2 Add/remove
+
+/// Remove items "item" to "item2" from list "l".
+///
+/// @warning Does not free the listitem or the value!
+///
+/// @param[out]  l  List to remove from.
+/// @param[in]  item  First item to remove.
+/// @param[in]  item2  Last item to remove.
+void tv_list_remove_items(list_T *const l, listitem_T *const item,
+                          listitem_T *const item2)
+  FUNC_ATTR_NONNULL_ALL
+{
+  // notify watchers
+  for (listitem_T *ip = item; ip != NULL; ip = ip->li_next) {
```

Comment:
One could replace `ip != NULL` with `ip != item2->li_next` and remove the `break` statement.